x/upgrade/keeper.go

```diff
@@ -1,97 +1,202 @@
+	store.Set(valAddress, VersionToBytes(version))
 }
 
-// GetUpgradedConsensusState get the expected upgraded consensus state for the next version of this chain
-func (k Keeper) GetUpgradedConsensusState(ctx sdk.Context, lastHeight int64) ([]byte, bool) {
+// DeleteValidatorVersion deletes a signalled version for a validator using the keeper's store key
+func (k Keeper) DeleteValidatorVersion(ctx sdk.Context, valAddress sdk.ValAddress) {
 	store := ctx.KVStore(k.storeKey)
-	bz := store.Get(types.UpgradedConsStateKey(lastHeight))
-	if len(bz) == 0 {
-		return nil, false
-	}
+	store.Delete(valAddress)
+}
 
-	return bz, true
+// EndBlock is called at the end of every block. It tallies the voting power that has
+// voted on each version. If one version has quorum, it is set as the quorum version
+// which the application can use as signal to upgrade to that version.
+func (k *Keeper) EndBlock(ctx sdk.Context) {
+	threshold := k.GetVotingPowerThreshold(ctx)
+	hasQuorum, version := k.TallyVotingPower(ctx, threshold.Int64())
+	if hasQuorum {
+		k.quorumVersion = version
+	}
 }
 
-// ClearIBCState clears any planned IBC state
-func (k Keeper) ClearIBCState(ctx sdk.Context, lastHeight int64) {
-	// delete IBC client and consensus state from store if this is IBC plan
+// TallyVotingPower tallies the voting power for each version and returns true if
+// any version has reached the quorum in voting power
+func (k Keeper) TallyVotingPower(ctx sdk.Context, threshold int64) (bool, uint64) {
+	output := make(map[uint64]int64)
 	store := ctx.KVStore(k.storeKey)
-	store.Delete(types.UpgradedClientKey(lastHeight))
-	store.Delete(types.UpgradedConsStateKey(lastHeight))
+	iterator := store.Iterator(nil, nil)
+	defer iterator.Close()
+	for ; iterator.Valid(); iterator.Next() {
+		valAddress := sdk.ValAddress(iterator.Key())
+		// check that the validator is still part of the bonded set
+		val, found := k.stakingKeeper.GetValidator(ctx, valAddress)
+		if !found {
+			// if it no longer exists, delete the version
+			k.DeleteValidatorVersion(ctx, valAddress)
+		}
+		// if the validator is not bonded, skip it's voting power
+		if !found || !val.IsBonded() {
+			continue
+		}
+		power := k.stakingKeeper.GetLastValidatorPower(ctx, valAddress)
+		version := VersionFromBytes(iterator.Value())
+		if _, ok := output[version]; !ok {
+			output[version] = power
+		} else {
+			output[version] += power
+		}
+		if output[version] > threshold {
+			return true, version
+		}
+	}
+	return false, 0
+}
+
+// GetVotingPowerThreshold returns the voting power threshold required to
+// upgrade to a new version. It converts the signal quorum parameter which
+// is a number between 0 and math.MaxUint32 representing a fraction and
+// then multiplies it by the total voting power
+func (k Keeper) GetVotingPowerThreshold(ctx sdk.Context) sdkmath.Int {
+	quorum := k.SignalQuorum(ctx)
+	totalVotingPower := k.stakingKeeper.GetLastTotalPower(ctx)
+	return quorum.MulInt(totalVotingPower).RoundInt()
```

Comment:
[blocking] I think it would be problematic if this rounds down and a quorum of less than 2 / 3 of validators is able to start the upgrade b/c the totalVotingPower is slightly less than 2/3... Should this always round up instead?

Reply:
As I said in the other comment, because the condition is strictly greater, this won't happen but you might get a case where it rounds up and with strictly greater ends up requiring more voting power than necessary so I'm going to change to using `Ceil()`

Reply:
Resolving b/c this seems outdated now.